Resolves #53: Implement quadratic symmetries and second moment twirl

poig opened the pull request:
This PR resolves #53 by implementing the `get_quadratic_symmetries` method and the general `second_moment` application function.

### Summary of Contribution

*   **New Method for Components:** Added a new method, ``get_graph_components`` with parameter of ``anticommutator`` or ``commutator``, to the `PauliStringCollection` class. This correctly computes the connected components of the **commutator** or **anticommutator** graph, as required for this feature.
*   **New Application Function:** Created the `second_moment(M, system)` function in `src/paulie/application/`. This function uses the new components method to correctly compute the second-order twirl for a general operator `M`.
*   **Supporting Methods:** Implemented and debugged necessary helper methods within the `PauliStringLinear` class (e.g., `__matmul__`, `__mul__`, `trace`, `simplify`, `__eq__`).
*   **Testing:** Added a comprehensive unit test suite (`tests/test_second_moment.py`) with multiple test cases, including a "zero case," a "positive case," and a test against the complex `a_5` system.
*   **Documentation:** Added a new documentation page, `docs/source/second_moment.rst`, explaining the theory and providing a clear usage example.

### Discussion and Implementation Choices

This implementation follows the general formula from Supplemental Theorem 1 of arXiv:2310.11505. After discovering that changing the existing `get_subgraphs` method would break other tests, a new `get_commutator_graph_components` method was created instead.

*   **This design choice is beneficial for flexibility and future development.** It allows different parts of the library to use different types of graph components (anti-commutation vs. commutator) as needed, providing more freedom to express different physical concepts without conflict.
*   As correctly proven by @Newtech66, the `Q_kj` basis is already orthogonal. Therefore, the original suggestion to use `get_dependents` was unnecessary and has been omitted.
*   The discussion also explored an advanced optimization for calculating `EZ`. This PR implements the general function as requested by the issue, leaving such optimizations for a potential future issue.

All tests in the new suite, as well as the library's existing tests, now pass, confirming that this is a complete and safe implementation of the requested feature.

gksmail commented:
1.
```python
        system = PauliStringCollection(p(["Z"]))
         M = PauliStringLinear([(1.0, "IZ")])

```
p(object) creates a PauliStringCollection if the object is a Pauli string array.

```python
        system = p(["Z"])
         M = p([(1.0, "IZ")])
    
```
Factory's idea to make working with different types of Pauli strings transparent
p("X") -> PauliString("X")
p(["X"]) -> PauliStringCollection(["X"]),
p([(1.0, "X")]) -> PauliStringLinear([(1.0, "X")]). 
This is done so that a person does not think about the class type, but thinks about the task.
In the future I want to develop a parser
p("X") - PauliString("X")
p("X,") - PauliStringCollection("X")
p("1.0*X") - PauliStringLinear((1.0, "X"))
I was thinking about introducing a special notation for this
p"X", p"X,", p"1.0*X"
This needs to be specifically adapted to Python grammar. I've abandoned this idea for now. But the meaning is something like f in print(f"{a}")

Therefore, it is better to construct through p, it will go through the parser.

gksmail commented:
2. You have included some code in the PauliStringCollection. The idea is right. For now, it is advisable to put all this in application. Since this is mixed with the decisions of other participants. Maybe we'll combine it later.
3.
```python
                    tensor_prod_str = str(S) + str(Lj_times_S)
                    linear_combination_terms.append((1.0, tensor_prod_str))
```
Here you don’t need to convert to strings and then construct PauliString
enough
`
```python
                    tensor_prod_str = S + Lj_times_S
                    linear_combination_terms.append((1.0, tensor_prod_str))
```
4. A linear basis is the single unconnected vertices in a graph. Therefore, there is no need to count them separately; the graph and subgraphs are enough.
5. Your output is PauliStringLinear, and where is the calculation of the moment itself. It's the same number.
6. You need to check the code on pylint, otherwise the workflow won't let you through.

`

poig commented:
> ```python
>         system = PauliStringCollection(p(["Z"]))
>          M = PauliStringLinear([(1.0, "IZ")])
> ```
> 
> p(object) creates a PauliStringCollection if the object is a Pauli string array.
> 
> ```python
>         system = p(["Z"])
>          M = p([(1.0, "IZ")])
>     
> ```
> 
> Factory's idea to make working with different types of Pauli strings transparent p("X") -> PauliString("X") p(["X"]) -> PauliStringCollection(["X"]), p([(1.0, "X")]) -> PauliStringLinear([(1.0, "X")]). This is done so that a person does not think about the class type, but thinks about the task. In the future I want to develop a parser p("X") - PauliString("X") p("X,") - PauliStringCollection("X") p("1.0_X") - PauliStringLinear([(1.0, "X")]) I was thinking about introducing a special notation for this p"X", p"X,", p"1.0_X" This needs to be specifically adapted to Python grammar. I've abandoned this idea for now. But the meaning is something like f in print(f"{a}")
> 
> Therefore, it is better to construct through p, it will go through the parser.

thank you for the feedback , I have now updated the unit test suite to use the ``p()`` factory for all object creation, removing the direct calls to the class constructors.

gksmail commented:
Look at the comments to the code above (2-6).

gksmail commented:
```python
        for linear_symmetry in linear_symmetries:
            for component_nodes in connected_components:
                linear_combination_terms = []
                for s_str in component_nodes:
                    s = self.create_instance(pauli_str=s_str)
                    linear_symmetry_times_s = linear_symmetry @ s
                    tensor_prod_str = str(s) + str(linear_symmetry_times_s)
                    linear_combination_terms.append((1.0, tensor_prod_str))

                if linear_combination_terms:
                    q_kj = PauliStringLinear(linear_combination_terms)
                    quadratic_basis.append(q_kj)

```

```python
 from paulie.common.pauli_string_factory import get_pauli_string as p

   pk = [p([(1, "II")])]
   pk.append(p([(1, "XY")]))
   pk.append(p([(1, "ZX")]))
   pk.append(p([(1, "YZ")]))
   pk.append(p([(1, "ZI"), (1, "XZ"), (1, "IX"), (1, "YY")]))
   pk.append(p([(1, "XI"), (1, "ZZ")]))
   pk.append(p([(1, "IY"), (1, "YX")]))
   pk.append(p([(1, "YI"), (1, "ZY")]))
   pk.append(p([(1, "IZ"), (1, "XX")]))

   lb = [p([(1, "II")])]
   lb.append(p([(1, "XY")]))
   lb.append(p([(1, "ZX")]))
   lb.append(p([(1, "YZ")]))
   
   qudratic = [k.quadratic(l[0][1]) for l in lb for k in pk]   
```
If you look closely at this example. Then pk contains the vertices of subgraphs, each with weight 1.
In lb - linear basis. 
 [k.quadratic(l[0][1]) for l in lb for k in pk] - returns a complete quadratic basis.
It seems to me that this is more local. I just didn't point this out in the example.
connected_components are subgraphs
```python

pk = p([(1, a) for a in s for s in connected_components])
lb = p([(1, a) for a in s for s in connected_components if len(s) == 1])

```

poig commented:
> ```python
>         for linear_symmetry in linear_symmetries:
>             for component_nodes in connected_components:
>                 linear_combination_terms = []
>                 for s_str in component_nodes:
>                     s = self.create_instance(pauli_str=s_str)
>                     linear_symmetry_times_s = linear_symmetry @ s
>                     tensor_prod_str = str(s) + str(linear_symmetry_times_s)
>                     linear_combination_terms.append((1.0, tensor_prod_str))
> 
>                 if linear_combination_terms:
>                     q_kj = PauliStringLinear(linear_combination_terms)
>                     quadratic_basis.append(q_kj)
> ```
> 
> ```python
>  from paulie.common.pauli_string_factory import get_pauli_string as p
> 
>    pk = [p([(1, "II")])]
>    pk.append(p([(1, "XY")]))
>    pk.append(p([(1, "ZX")]))
>    pk.append(p([(1, "YZ")]))
>    pk.append(p([(1, "ZI"), (1, "XZ"), (1, "IX"), (1, "YY")]))
>    pk.append(p([(1, "XI"), (1, "ZZ")]))
>    pk.append(p([(1, "IY"), (1, "YX")]))
>    pk.append(p([(1, "YI"), (1, "ZY")]))
>    pk.append(p([(1, "IZ"), (1, "XX")]))
> 
>    lb = [p([(1, "II")])]
>    lb.append(p([(1, "XY")]))
>    lb.append(p([(1, "ZX")]))
>    lb.append(p([(1, "YZ")]))
>    
>    qudratic = [k.quadratic(l[0][1]) for l in lb for k in pk]   
> ```
> 
> If you look closely at this example. Then pk contains the vertices of subgraphs, each with weight 1. In lb - linear basis. [k.quadratic(l[0][1]) for l in lb for k in pk] - returns a complete quadratic basis. It seems to me that this is more local. I just didn't point this out in the example. connected_components are subgraphs
> 
> ```python
> pk = p([(1, a) for a in s for s in connected_components])
> lb = p([(1, a) for a in s for s in connected_components if len(s) == 1])
> ```

Updated with your advice, thank you for the detailed follow-up comments and for providing the code examples. They were very helpful and prompted a much deeper investigation into the library's structure.

I spent some time trying to refactor my implementation to use the existing .quadratic() method as you suggested. Unfortunately, I was unable to get the correct final result for second_moment when using it. The basis of Q_kj vectors it produced seemed to be different from what the projection formula requires, leading to mathematically incorrect results in the unit tests (e.g., second_moment("IZ") did not return "IZ").

After this investigation, I have reverted to my initial implementation. This version follows the formula Q_kj = ∑_{S ∈ C_k} S ⊗ (L_j*S) from the paper directly. It is a bit more verbose, but it is self-contained and demonstrably produces the correct physical results in the test suite.

I believe that the existing .quadratic() method may be intended for a different application or may have a subtle bug, which is a separate issue that could be addressed in the future.

gksmail commented:
I asked @AmanieOxana to prepare independent test cases. 
There are lint errors
https://github.com/QPauLie/PauLie/actions/runs/15497592266/job/43638544136?pr=62

poig commented:
> I asked @AmanieOxana to prepare independent test cases. There are lint errors https://github.com/QPauLie/PauLie/actions/runs/15497592266/job/43638544136?pr=62

done! if everything looks ok I think its ready to merge

Newtech66 commented:
@poig Your code fails to reproduce [Supplemental Example 2](https://arxiv.org/pdf/2310.11505). The quadratic symmetries of `p(["X"])` are

```bash
II
IX
ZZ + YY
i*YZ - i*ZY
XX
XI
```
however your code outputs
```bash
II
IX
ZZ + YY
ZY + YZ
XX
XI
```
As you can see, it has the incorrect sign for `ZY` (ignoring the complex factor).

poig commented:
> @poig Your code fails to reproduce [Supplemental Example 2](https://arxiv.org/pdf/2310.11505). The quadratic symmetries of `p(["X"])` are
> 
> ```shell
> II
> IX
> ZZ + YY
> i*YZ - i*ZY
> XX
> XI
> ```
> 
> however your code outputs
> 
> ```shell
> II
> IX
> ZZ + YY
> ZY + YZ
> XX
> XI
> ```
> 
> As you can see, it has the incorrect sign for `ZY` (ignoring the complex factor).

corrected

gksmail commented:
@poing use @  multiplication - it only returns the ~ Pauli string. We didn't need a normal piece before. This is a proportional product. 
@ - do not change this operator. It is needed in other algorithms.
When doing a linear combination, it was necessary to do a normal product. So I quickly entered the meaning of the sign. This is a temporary solution.
```python
    def quadratic(self, basis:PauliString):
        """
        Quadratic form
        Returns a linera comination of PauliString
        """
        new_combinations = []
        for c in self.combinations:
            new_combinations.append((c[0]*basis.sign(c[1]), c[1] + c[1]@basis))
        return PauliStringLinear(new_combinations)

```

I suggest testing this function and if there is anything to fix.

Newtech66 commented:
The sign method needs correction in this PR:

```python
from paulie.common.pauli_string_factory import get_pauli_string as p

print(p("XX").sign(p("XY"))
```
outputs `1`, when it should in fact be `i`.

Newtech66 commented:
The following example is incorrect:

```python
from paulie.common.pauli_string_factory import get_pauli_string as p
from paulie.application.second_moment import second_moment, get_full_quadratic_basis

g = p(["X"])
print('Quadratic symmetries:')
print(*get_full_quadratic_basis(g), sep='\n')

op1 = p([(1, "IX")])
print('Second-order twirl of IX:', second_moment(op1, g))

op2 = p([(1, "YY")])
print('Second-order twirl of YY:', second_moment(op2, g))

op3 = p([(1, "YZ"), (-1, "ZY")])
print('Second-order twirl of YZ - ZY:', second_moment(op3, g))
```

which outputs

```bash
Quadratic symmetries:
II
IX
YY + ZZ
i*YZ - i*ZY
XX
XI
Second-order twirl of IX: IX
Second-order twirl of YY: 0.5*YY + 0.5*ZZ
Second-order twirl of YZ - ZY: YZ + ZY
```
`YZ - ZY` is a quadratic symmetry, therefore its twirl should also be `YZ - ZY`, right?

gksmail commented:
> > Thanks. I'll take a look now.
> 
> My PR for this issue has a corrected `sign` function, you can take a look.

fixed

poig commented:
I think mine might be better in some cases compare to the original loop method, as it can be slow

poig commented:
@Newtech66 did you catch any other error need to be fix? i am happy to take a deeper look into them

Newtech66 commented:
@poig, could you try the OTOC tests from my PR #64 for this issue? They are quite strong, and should be able to catch errors. You don't have to merge those tests into this PR, because it's dependent on any PR for the OTOC issue (#57 or #65).

Newtech66 commented:
@poig, is the example from [my comment above](https://github.com/QPauLie/PauLie/pull/62#issuecomment-2952277903) passing yet?

Newtech66 commented:
@poig, I see you have used the `unittest` framework for writing test cases for this PR. For uniformity, I would suggest you write the tests using `pytest` instead, to maintain uniformity with the rest of the repository. @gksmail, do you agree?

gksmail commented:
@poig fix lint  
https://github.com/QPauLie/PauLie/actions/runs/15506872075/job/43663412893?pr=62

poig commented:
@Newtech66 yes you are right, I am fixing it and other error on this library leading the error

poig commented:
/

> > Ok, I cross-checked this implementation with the OTOC tests, and they pass. However, it still doesn't pass the `YZ - ZY` example. This means the OTOC tests aren't as strong as I had hoped. And that there is still something wrong with the implementation in this PR.
> 
> I propose to deal with this PR now. Make a merge. And deal with your test problem.

I agree as two pr come together will make a lot of mess

poig commented:
@Newtech66 fixed check again

Newtech66 commented:
The relative sign between `YZ` and `ZY` is now correct. However, there are other issues:

- You output one of the quadratic symmetries as `i*YZ - i*ZY`. If you use this convention, then the twirl should be `ZY - YZ` and not `YZ - ZY`. But this point may just be a nitpick.
- As I understand, the function `p(["X"]).get_full_quadratic_basis(normalized=True)` outputs the orthonormalized quadratic basis. However, the output is not correct: everything needs to be divided by 2. For example, you output `II`, but `trace(II *II) = 4`, which is clearly not normalized. The normalized operator should be `0.5*II`.

poig commented:
> The relative sign between `YZ` and `ZY` is now correct. However, there are other issues:
> 
> * You output one of the quadratic symmetries as `i*YZ - i*ZY`. If you use this convention, then the twirl should be `ZY - YZ` and not `YZ - ZY`. But this point may just be a nitpick.
> * As I understand, the function `p(["X"]).get_full_quadratic_basis(normalized=True)` outputs the orthonormalized quadratic basis. However, the output is not correct: everything needs to be divided by 2. For example, you output `II`, but `trace(II *II) = 4`, which is clearly not normalized. The normalized operator should be `0.5*II`.


for normalization you are right since The correctly normalized operator is 0.35355 * (YY + ZZ). My code outputs 0.707... * (YY + ZZ), which is off by a factor of sqrt(2).
, but 
``T(M) = [Tr(Q†M) / Tr(Q†Q)] * Q.``
if 
- M = YZ - ZY
- Q = i*YZ - i*ZY

then 
Q† = (i*YZ - i*ZY)†
= (i*YZ)† - (i*ZY)†
= (-i)*(Z†Y†) + (i)*(Y†Z†) (since (AB)† = B†A†) and (c*A)† = c* * A†)
= (c*A)† = c* * A†)
Q† = i*YZ - i*ZY
Q† = -Q

Tr(Q†Q) = Tr( (-Q) * Q ) = -Tr(Q²).
The easiest way is to note the relationship M = -i * Q
Tr(Q†M) = Tr( (-Q) * (-i*Q) ) = Tr( i * Q² ) = i * Tr(Q²)

coeff = Numerator / Denominator = (i * Tr(Q²)) / (-Tr(Q²))
coeff = -i.

T(M) = coeff * Q = (-i) * Q = (-i) * (i*YZ - i*ZY)
T(M) = (-i²)*YZ - (-i²)*ZY = 1*YZ - 1*ZY = YZ - ZY.

Therefore, twirl $ T(YZ - ZY) = YZ - ZY 

am I right?

Newtech66 commented:
> = (iYZ)† - (iZY)†
= (-i)(Z†Y†) + (i)(Y†Z†) (since (AB)† = B†A†) and (cA)† = c * A†)

That's incorrect: `ZY` is the **tensor product** $Z \otimes Y$. Therefore, $(Z \otimes Y)^{\dagger} = $Z^{\dagger} \otimes Y^{\dagger} = Z \otimes Y$.

poig commented:
> > = (iYZ)† - (iZY)†
> > = (-i)(Z†Y†) + (i)(Y†Z†) (since (AB)† = B†A†) and (cA)† = c * A†)
> 
> That's incorrect: `ZY` is the **tensor product** Z ⊗ Y . Therefore, ( Z ⊗ Y ) † = Z † ⊗ Y † = Z ⊗ Y .

it might still be the sign function problem, since I seperate all function in small and 100% sure already, but I rethink about it
You’re absolutely right to point out the mistake in the Hermitian conjugate calculation. I interpreted ZY as a product of operators on the same Hilbert space, applying $(AB)† = B†A†$, but in the context of multi-qubit systems, ZY represents the tensor product Z ⊗ Y. For tensor products, the correct rule is $(A ⊗ B)† = A† ⊗ B†$, and since Pauli matrices are Hermitian ($ Z† = Z $, $ Y† = Y $), we have:

$(Z ⊗ Y)† = Z† ⊗ Y† = Z ⊗ Y$

Let’s correct the calculation for $ Q = i*YZ - i*ZY $, assuming $ YZ = Y ⊗ Z $ and $ ZY = Z ⊗ Y $:

$Q = i*(Y ⊗ Z) - i*(Z ⊗ Y)$
$Q† = [i*(Y ⊗ Z)]† + [-i*(Z ⊗ Y)]†$
$[i*(Y ⊗ Z)]† = (-i)*(Y ⊗ Z)† = (-i)*(Y ⊗ Z)$
$[-i*(Z ⊗ Y)]† = (i)*(Z ⊗ Y)† = (i)*(Z ⊗ Y)$
So, $ Q† = (-i)*(Y ⊗ Z) + (i)*(Z ⊗ Y) = -Q$

This shows $Q$ is anti-Hermitian, not Hermitian as the original calculation suggested. The step $(-i)*(Z†Y†) + (i)*(Y†Z†)$ was incorrect because it treated ZY as a single-space product rather than a tensor product.

am I right?

Newtech66 commented:
You're right. I was mistaken.

poig commented:
@gksmail kindly remind that the unitary hack need to "assign people to the issues" and "close issues" to be consider as complete the bounties before June 11th 5pm, and I won't be available for the next few days.

Newtech66 commented:
> unitary hack need to "assign one person to the issues"

I think exceptional cases where the maintainers combine parts from two different pull requests are also possible. @vprusso may have some input on this. I remember such a situation with toqito from a previous UnitaryHack.

gksmail commented:
@poig

poig commented:
> @poig makes rebase

sorry, is there anything need help to merge?

gksmail commented:
@poig and @vprusso 


> @gksmail I am just kindly remind that the unitary hack need to "assign one person to the issues" and "close issues" to be consider as complete the bounties by the script before June 11th 5pm, even though not sure we should unite that but I accept anything that's fair from you two because I won't be available for the next few days, Thank you.
I asked the same question yesterday @vprusso. 
After talking with the organizers, he answered me;
> Hiya! We said we wouldn't split bounties this year, I think we should stick to that. Not sure if the hackers have interacted with one another at all, but if yes they could consider teaming up. Although that still means one person gets the bounty and they'd have to split it amongst themselves
According to the rules, we can only accept one PR per task. 
My position is that working together on a problem is better. Since we are aimed at people who are interested in the project, who want to develop it. And who is ready to work together.
My decision to accept #62 PR is more political. @Newtech66  did a great job on decomposition and you are already a contributor to the project. I would like @poig to be able to document his work as a contributor.

poig commented:
> > > @poig makes rebase
> > 
> > 
> > is there anything need help to merge?
> 
> You changed the sign function, and it now conflicts with main branch. You need to do git rabase original main. And resolve conflicts.

is it ok now?

gksmail commented:
![image](https://github.com/user-attachments/assets/0bc51601-58fb-450c-87d4-bac9e49e1628)

poig commented:
![image](https://github.com/user-attachments/assets/fbb413d1-586b-4412-bc03-86d40dd63098)

poig commented:
@gksmail is it resolve from your side?

gksmail commented:
![image](https://github.com/user-attachments/assets/6002009e-d285-496c-921a-f74633987c5b)
no

poig commented:
> ![image](https://private-user-images.githubusercontent.com/22450477/452721436-6002009e-d285-496c-921a-f74633987c5b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDkzNjMzNDEsIm5iZiI6MTc0OTM2MzA0MSwicGF0aCI6Ii8yMjQ1MDQ3Ny80NTI3MjE0MzYtNjAwMjAwOWUtZDI4NS00OTZjLTkyMWEtZjc0NjMzOTg3YzViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA2MDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNjA4VDA2MTA0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzZDExNzNhZDA4NGJkMmY3MjNkOWMwMGIyM2VmMjQwZDg0MWEyMmU0YjI0NzBlNjQwYjNlYzc2YTI0NDRkZjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.inO1LmpNO4P3tTLIXSYpIEObig0i9fnhAzA6IvcGAss) no

hmm it seems like its your side need to do something then

poig commented:
I will make a new pr, and see how it goes
